### ISSUE TITLE
Test functions called by module itself

### ISSUE THREAD

**Issue by JuPrgn** [ThrowTheSwitch/Ceedling#604]
Hi,

I have a question on the best way to test functions called by the module under test itself. This is a recurring problem for me because most test units call setter-type functions or static functions.
Let me introduce a simple example : 

`Sensor.c` : 
```
#include "Sensor.h"
#include "Error.h"

STATIC READ_MODES_t mode;
STATIC int16_t offset;

int16_t ConfigureParameters(READ_MODES_t Mode, int16_t Offset)
{
    // Set mode
    int16_t ret = SetMode(Mode);
    if (ERROR_IsError((Ret)))
    {
         return (Ret);
    }  

   // Set offset
    ret = SetOffset(Offset);
    if (ERROR_IsError((Ret)))
    {
         return (Ret);
    }  
    return ERROR_GetCode(ERR, NO_ERROR);
}

int16_t SetMode(READ_MODES_t Mode)
{
    switch (Mode)
    {
        case READ_MODE_UNIDIRECTIONAL:
            mode = READ_MODE_UNIDIRECTIONAL;
            break;
        case READ_MODE_BIDIRECTIONAL:
            mode = READ_MODE_BIDIRECTIONAL;
            break;
        default:
            return ERROR_GetCode(ERR, INVALID_MODE);
            break;
    }
    return ERROR_GetCode(ERR, NO_ERROR);
}

int16_t SetOffset(int16_t Offset)
{
    if ((Offset < 0) || (Offset >= MAX_VALUE))
    {
        return ERROR_GetCode(ERR, OFFSET_OUT_OF_RANGE);
    }
    offset = Offset;
    return ERROR_GetCode(ERR, NO_ERROR);
}
```

With `test_Sensor.c` : 
```
#include "unity.h"

#include "Sensor.h"
#include "mock_Error.h"

extern READ_MODES_t mode;
extern int16_t offset;

void setUp(void)
{
}

void tearDown(void)
{
}

void test_Sensor_ConfigureParameters(void)
{
    mode = READ_MODE_BIDIRECTIONAL;
    offset = 200;
    ERROR_GetCode_ExpectAndReturn(ERR, NO_ERROR, 0);// From SetMode call
    ERROR_IsError_ExpectAndReturn(0, false);
    ERROR_GetCode_ExpectAndReturn(ERR, NO_ERROR, 0);// From SetOffset call
    ERROR_IsError_ExpectAndReturn(0, false);
    ERROR_GetCode_ExpectAndReturn(ERR, NO_ERROR, 0);// From ConfigureParameters
    PS_ConfigureParameters(READ_MODE_UNIDIRECTIONAL, 30);
    TEST_ASSERT_TRUE(READ_MODE_UNIDIRECTIONAL == mode);
    TEST_ASSERT_EQUAL_INT16(30, offset);
}
```
It becomes very complicated when several functions are called or when functions themselves call other functions. I would like to use something simple working as mock to only check if SetMode function has been called... This also apply for static functions that do lower abstraction things this is sometimes hard to check if a function has been called in a test knowing that the static function itself is tested in another test. I would share some other example in a second time.

I have read on other issues that a solution was to create a helper module but this requires the use of a lot of files if an helper function call another helper function and does not seem to me to be the best solution. 
Do you have any other ideas for testing this kind of code?

**Comment by Letme** [comment]
We have a bit another way around, but I can revert our solution to yours.

Static functions are not declared in header files -> if you declare them in a separate header file (let's call it `Sensor_private.h` based on your example), you can simply include that file in the `Sensor.c` and compiler/linker will not complain. However, it will allow you to call `mock_Sensor_private.h` in your unit test. I don't know if that is our build system or if that is Ceedling (we don't use Ceedling), but this way you can mock static functions from the module itself. Be careful that you afterwards, in the split test (separate compilation unit) to also test the unmocked version.

**Comment by JuPrgn** [comment]
@Letme Thank you for your answer.

If I got it well, I would have : 

- Sensor.c
- Sensor.h
- Sensor_private.h
- test_Sensor.c -> Test Sensor.h functions mocking private header
- test_Sensor_private.c -> Test static functions (unmocked version)

**Comment by Letme** [comment]
Yes, and `Sensor.c` would include both `Sensor.h` and `Sensor_private.h` so that compiler can tell you in case you mismatch some declaration/definitions in the future development.

**Comment by JuPrgn** [comment]
OK I will give it a try !
And for functions that are not static as SetOffset that may be call by another unit and also by sensor is there a trick ? Or should I put them on Sensor_private.h and include Sensor.h and Sensor_private.h on those other units that use them ?

**Comment by Letme** [comment]
You are pushing too much - what else will then remain as a real implementation to test? You do not want to mock everything, otherwise just mock `Sensor.h`. External units may never include `Sensor_private.h` as those functions are not part of the unit's API. Since I assume `Sensor_private.h` might have to include `Sensor.h` for some typedefs, then duplicate functions will be the problem. And external units will never require details of `Sensor_private.h` as `mock_Sensor.h` will prevent all calls to the further functions anyway.

**Comment by JuPrgn** [comment]
I will try some code to see more clearly. 
Thanks for your advices !

**Comment by JuPrgn** [comment]
It works fine running `test_Sensor_private.c` but linker complains about my static functions being multiple defined  whenever I use `mock_Sensor.h` on `test_Sensor.c`. 
I think I should move functions to `Sensor_private.c` to fix this ? Do you keep static functions on your `Sensor.c` without issues ?

**Comment by JuPrgn** [comment]
I kept them defined as STATIC on _private.h, using extern keyword fix linker messages

**Comment by JuPrgn** [comment]
I am probably using it wrong because even if I can include `mock_Sensor_private.h` mock doesn't output anything for those static functions (mock_Sensor_private.h and .c are almost empty) and then I get undefined reference to `_static_function_ExpectAndReturn`.

**Comment by Letme** [comment]
So you have something like:

`Sensor.h`:
```
void funcB(int c);
```
`Sensor_private.h` (don't forget ifdef guards):
```
#include "Sensor.h" // Maybe needed for typedefs

STATIC void funcA(int a, int b);
```

`Sensor.c`:
```
#include Sensor.h
#include Sensor_private.h

STATIC void funcA(int a, int b)
{
    (void) a;
    (void) b;
}

void funcB(int c)
{
    funcA(c, 10);
}
```

`Test_Sensor.c`:
```
#include Sensor.h
#include mock_Sensor_private.h

void test_funcB(void)
{
    FuncA_Expect(5,10);
    funcB(5);
}
```

`project.yml` has define `STATIC=static` for release, test define is `STATIC=''`

**Comment by JuPrgn** [comment]
Yes this is exactly what I have but I get linker : multiple definition of FuncA.

So I changed `Sensor_private.h` to 

```
#include "Sensor.h" // Maybe needed for typedefs

extern void funcA(int a, int b);
```
This remove the linker error but I get no `_Expect` function on `mock_Sensor_private.c` and .h

**Comment by Letme** [comment]
In your example even extern is useless in header files. Its not a variable so any non static label is extern.

Where does it find multiple definitions? Maybe mock and real? You have `--allow-multiple-definitions` flag to linker and then it takes the first function it finds (include order matters), so that might be more of a quick fix.

**Comment by JuPrgn** [comment]
> Where does it find multiple definitions? Maybe mock and real? You have `--allow-multiple-definitions` flag to linker and then it takes the first function it finds (include order matters), so that might be more of a quick fix.

How/Where do you add this flag ? I get error at link time using this flag (with and without plurial `definition`) : 

> gcc: error: unrecognized command line option ‘--allow-multiple-definition’; did you mean ‘-Wold-style-definition’?

This is my specific GCC project.yml : 
```
---
# Specific configuration to run on Linux using GCC
# these options are added to default project.yml
#
# Required : Install GCOV and GCOVR
#
:defines:
  :test:
    - TEST_GCC
  :test_preprocess:
    - TEST_GCC
    
:flags:
  :test:
    :compile:
      :*:
        - -fpack-struct
    :link:
      :*:
        - --allow-multiple-definition
  :release:
    :compile:
      :*:
        - -fpack-struct
  :gcov:
    :compile:
      :*:
        - -fpack-struct

:tools_test_linker:
  :arguments:
    - -lm # Required for <math.h>
    
:tools_gcov_linker:
  :arguments:
    - -lm # Required for <math.h>
    
:plugins:
  :enabled:
    - gcov
...
```

**Comment by Letme** [comment]
I think it needs to be like in the test:linker:flags:: `-Wl,--allow-multiple-definition`

**Comment by JuPrgn** [comment]
Thank you this option give warning instead of error. I have test_linker on XC16.yml because I had to redefine executable... but I need to figure out if I can do the same with GCC.yml without redefining every options.

This is how you are using it too in your projects ?

**Comment by Letme** [comment]
Like I said. In simpler projects, where I am using Ceedling I never needed to mock internal static functions. And I am opposed to putting definitions of `static inline` functions into the header files, so I avoid hitting also similar problems to yours. Your use case is kinda niche. The fact is, that what you have in `_private`, it should be whole new compilation unit. As you are probably sharing the global variables with others in `Sensor.c` then its not actually an option to move. So this kind of architecture is bad for testing.

As for multiple definition usage I have a big system, using normal Makefile build system (but using CMock). Because of the static inline functions in header files there are quite some problems presented, one of these is such multiple definition problem as functions found in mock are also found in real implementation. So to solve this you first include mocks and then include the real implementation (which means you might also have a use case where you would call mocks while you might think you are calling real implementation and that is a risk). To make a long story short, what you are pasting me is normal verbosity, so I suggest you change the usual `ceedling` call to `ceedling verbosity[4]` (check the docs for exact command, as I know its something weird comparing to other targets).

**Comment by JuPrgn** [comment]
Okay, it's the architecture that is not suitable in my case. 
I saw the architectures based on model conductor hardware (and the use case in Ceedling example) it seems that it avoids a lot of problems and simplifies the tests but I have a lot of existing code which is not based on this principle and I tried to find an architecture more close to what I currently use. Anyway MCH will probably be the next improvment step.

I understood that it is risky to use several definitions and I will try to break down my modules better to avoid this type of trick.
